### PR TITLE
[Bug fix] Fit SDPA decode correction in FP32 half-sync DST

### LINKED ISSUE #56171: scaled_dot_product_attention_decode produces wrong output with k_chunk_size=32 once a second chunk is in range
`ttnn.transformer.scaled_dot_product_attention_decode` returns garbage whenever the KV chunk size
is 32 and `cur_pos >= 32`, i.e. as soon as attention spans more than one chunk. Output is correct
while only the first chunk is attended, and correct for chunk sizes 64 and 128 with any number of
chunks (2, 3, 4 and 8 tested). Seen with head sizes 32 and 128, batch 1, on Wormhole (T3K,
`tt-metal` `058a20e4941c`). Not tested on Blackhole.

The kernel's own default hits this: with no `program_config`, `get_chunk_size()`
(`ttnn/cpp/ttnn/operations/transformer/sdpa_decode/sdpa_decode.cpp:22`) picks the largest
power-of-two divisor of the KV length, which is 32 for any KV length of the form 32·odd (96, 160,
224, …). So a plain call with a 96-position cache is wrong out of the box.

Found while running FIBO-vlm (Qwen3-VL) decode through `models/tt_dit/encoders/transformer.py`:
a 96-position KV cache with 32-position chunks produced garbage tokens, while the same model with
a 128-position cache and one 128 chunk decodes correctly.

**Reproduction:**

```python
import torch
import ttnn

heads, kv_heads, head_dim, cache = 8, 2, 128, 128
device = ttnn.open_device(device_id=0)
torch.manual_seed(0)
q = torch.randn(1, 1, heads, head_dim)
k = torch.randn(1, kv_heads, cache, head_dim)
v = torch.randn(1, kv_heads, cache, head_dim)
tq, tk, tv = (ttnn.from_torch(t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device) for t in (q, k, v))

for chunk in (32, 64):
    for cur_pos in (31, 32, 64):
        out = ttnn.transformer.scaled_dot_product_attention_decode(
            tq, tk, tv, cur_pos=[cur_pos], is_causal=True,
            program_config=ttnn.SDPAProgramConfig(
                compute_with_storage_grid_size=device.compute_with_storage_grid_size(),
                q_chunk_size=32, k_chunk_size=chunk, exp_approx_mode=False,
            ),
            compute_kernel_config=ttnn.WormholeComputeKernelConfig(
                math_fidelity=ttnn.MathFidelity.HiFi4, math_approx_mode=False, fp32_dest_acc_en=True, packer_l1_acc=False,
            ),
        )
        valid = cur_pos + 1
        kr = k[:, :, :valid].repeat_interleave(heads // kv_heads, dim=1)
        vr = v[:, :, :valid].repeat_interleave(heads // kv_heads, dim=1)
        ref = (torch.softmax(q.permute(0, 2, 1, 3) @ kr.transpose(-1, -2) / head_dim**0.5, dim=-1) @ vr).permute(0, 2, 1, 3)
        x, y = ref.double().flatten(), ttnn.to_torch(out)[:, :, :heads].double().flatten()
        x, y = x - x.mean(), y - y.mean()
        print(f"k_chunk_size={chunk} cur_pos={cur_pos}: PCC {(x @ y / (x.norm() * y.norm())).item() * 100:.2f} %")

ttnn.close_device(device)
```

**Diagnostics / observation:**

```
k_chunk_size=32 cur_pos=31: PCC 99.98 %
k_chunk_size=32 cur_pos=32: PCC 32.82 %
k_chunk_size=32 cur_pos=64: PCC 22.59 %
k_chunk_size=64 cur_pos=31: PCC 99.98 %
k_chunk_size=64 cur_pos=32: PCC 99.98 %
k_chunk_size=64 cur_pos=64: PCC 99.97 %
```

Full sweep (cache 96/128/256, `cur_pos` 0…cache−1, head sizes 32 and 128): every chunk-32 case
is at 100 % for `cur_pos < 32` and between 3 % and 88 % for `cur_pos >= 32`, worst exactly at
chunk boundaries (`cur_pos` 32, 64). Every chunk-64 and chunk-128 case is above 99.97 %.

### PR DESCRIPTION
**Pull request by pavlejosipovic**
### PR Category

Bug fix

### Summary

Fixes #56171.

SDPA decode's cross-core softmax correction loads five destination tiles, but FP32 accumulation with half synchronization only provides four slots. On Blackhole QB2 with `k_chunk_size=32`, the issue's repro drops from 99.98% PCC at position 31 to 32.82% at position 32, when the second core starts contributing. Device assertions identify the out-of-bounds copy into slot 4.

The five-slot correction was introduced by [3fc9fff9d0b0cab0c205c2928afff345052d8a23](https://github.com/tenstorrent/tt-metal/commit/3fc9fff9d0b0cab0c205c2928afff345052d8a23) in #32501, **SDPA Decode Optimization: Fused softmax correction LLK**. #34133 subsequently moved the helper into the shared SDPA compute header.

For FP32 half-sync, reuse slot 2: read `worker_sum` before overwriting it with `cur_max`, retain the corrected worker sum in an SFPU register, and write the combined sum to slot 3. The caller's synchronization mode is preserved. Other modes keep the existing five-slot implementation and rounding behavior.

Keep one focused decode regression: 32-token KV chunks, two cores per KV head, FP32 accumulation, half sync, and positions 31/32. Extend the existing LLK tests to exercise the four-slot layout. The golden change adapts tile layout only; the softmax equations and tolerances are unchanged.

The shared and Quasar wrappers select the corrected layout and assert DST capacity. The BGE-M3 model-local copy is excluded from this PR; its currently uncalled correction helper is deferred.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:pjosipovic/fix-sdpa-decode-fp32-half-sync)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:pjosipovic/fix-sdpa-decode-fp32-half-sync)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:pjosipovic/fix-sdpa-decode-fp32-half-sync)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:pjosipovic/fix-sdpa-decode-fp32-half-sync)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:pjosipovic/fix-sdpa-decode-fp32-half-sync)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:pjosipovic/fix-sdpa-decode-fp32-half-sync)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:pjosipovic/fix-sdpa-decode-fp32-half-sync)

### CODE REVIEW

**Review by skrsticTT — APPROVED** [review]

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/experimental/llk_sfpu/ckernel_sfpu_sdpa.h [hunk 1]
```diff
@@ -265,11 +268,20 @@ inline void calculate_fused_max_sub_exp_add_tile(int scale_bf16) {
...
-        sfpi::dst_reg[prev_sum_base_idx] = exp_prev * prev_sum_vec;
-        sfpi::vFloat corr_worker_sum = sfpi::dst_reg[worker_sum_base_idx];
-        sfpi::vFloat corr_prev_sum = sfpi::dst_reg[prev_sum_base_idx];
-        sfpi::dst_reg[prev_sum_base_idx] = corr_worker_sum + corr_prev_sum;
+        if constexpr (reuse_cur_max_tile) {
+            // Keep the corrected worker sum in an SFPU register: its input DST
+            // tile now holds cur_max, which must survive for the caller to pack.
+            sfpi::vFloat corr_worker_sum = exp_worker * worker_sum_vec;
+            sfpi::dst_reg[prev_sum_base_idx] = exp_prev * prev_sum_vec;
+            sfpi::vFloat corr_prev_sum = sfpi::dst_reg[prev_sum_base_idx];
```
Comment (nvelickovicTT) [inline tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/experimental/llk_sfpu/ckernel_sfpu_sdpa.h:276]: nit: some of the lines are common for both if and else clause. It would be great for readability if they are hoisted. And it is easier to see what is actually different in the 2 modes.

**Review by nvelickovicTT — APPROVED** [review]

**Review by nsorabaTT — APPROVED** [review]
